core/src/main/scala/cats/mtl/Handle.scala

```diff
@@ -218,5 +219,40 @@ private[mtl] trait HandleInstances extends HandleLowPriorityInstances {
 }
 
 object Handle extends HandleInstances {
+
   def apply[F[_], E](implicit ev: Handle[F, E]): Handle[F, E] = ev
+
+  def ensure[F[_], E]: AdHocSyntax[F, E] =
+    new AdHocSyntax[F, E]
+
+  final class AdHocSyntax[F[_], E] {
+
+    def apply[A](body: Handle[F, E] => F[A])(implicit F: ApplicativeThrow[F]): Inner[A] =
+      new Inner(body)
+
+    final class Inner[A](body: Handle[F, E] => F[A])(implicit F: ApplicativeThrow[F]) {
+      def recover(h: E => F[A]): F[A] = {
+        val Marker = new AnyRef
+
+        def inner[B](fb: F[B])(f: E => F[B]): F[B] =
+          ApplicativeThrow[F].handleErrorWith(fb) {
+            case Submarine(e, Marker) => f(e.asInstanceOf[E])
+            case t => ApplicativeThrow[F].raiseError(t)
+          }
+
+        val fa = body(new Handle[F, E] {
+          def applicative = Applicative[F]
+          def raise[E2 <: E, B](e: E2): F[B] =
+            ApplicativeThrow[F].raiseError(Submarine(e, Marker))
+          def handleWith[B](fb: F[B])(f: E => F[B]): F[B] = inner(fb)(f)
+        })
+
+        inner(fa)(h)
+      }
+    }
+  }
+
+  private final case class Submarine[E](e: E, marker: AnyRef)
+      extends RuntimeException
```

Comment:
Should this just extend `Throwable?  (Or even `ControlThrowable`, but I think that one is considered fatal by `NonFatal`.) It's not really an ordinary exception, and even if it leaked, without a stacktrace it would be hard to do something about it 😅

Reply:
I went back and forth. Technically RuntimeException is accurate and appropriate, but it's all sort of dependent on perspective. Certainly anything that would be considered fatal is off the table since it just instantly torpedoes the whole runtime.